File: evals/fixtures/python-lib/src/birdlog/store.py
"""The observation store: one JSON line per observation in a single file."""

import json
import os
from dataclasses import asdict
from pathlib import Path

from birdlog.models import Observation

# Maximum number of observations the store holds; log() refuses beyond this.
MAX_OBSERVATIONS = 5000

# Filename of the default store, inside the birdlog directory.
DEFAULT_FILENAME = "sightings.jsonl"

# Directory the default store lives in, under the user's home.
DEFAULT_DIR = ".birdlog"


class Store:
    """A JSONL-backed store of observations."""

    def __init__(self, path: Path) -> None:
        self.path = path

    @classmethod
    def open_default(cls) -> "Store":
        """Opens the store under $BIRDLOG_HOME or ~/DEFAULT_DIR."""
        base = Path(os.environ.get("BIRDLOG_HOME", str(Path.home() / DEFAULT_DIR)))
        return cls(base / DEFAULT_FILENAME)

    def log(self, observation: Observation) -> None:
        """Appends one observation; refuses once MAX_OBSERVATIONS is reached."""
        observations = self.read_all()
        if len(observations) >= MAX_OBSERVATIONS:
            raise ValueError(f"store is full ({MAX_OBSERVATIONS} observations)")
        if observations and observations[-1] == observation:
            return  # 1.4.2: never store the same sighting twice in a row
        observations.append(observation)
        self.write_all(observations)

    def read_all(self) -> list[Observation]:
        """Reads every observation, oldest first."""
        if not self.path.exists():
            return []
        return [
            Observation(**json.loads(line))
            for line in self.path.read_text().splitlines()
        ]

    def write_all(self, observations: list[Observation]) -> None:
        """Rewrites the store file, one JSON line per observation."""
        self.path.write_text(
            "".join(json.dumps(asdict(o)) + "\n" for o in observations)
        )

File: evals/fixtures/rust-cli/src/store.rs
//! The note store: one JSON line per note in a single store file.

use std::path::PathBuf;

/// Maximum number of notes the store holds; `add_note` refuses beyond this.
pub const NOTE_LIMIT: usize = 1000;

/// Directory the default store lives in, under the user's home.
pub const DEFAULT_DIR: &str = ".notes";

// TODO: export to markdown

pub struct Store {
    dir: PathBuf,
}

impl Store {
    /// Opens the store under `$NOTECLI_HOME` or `~/DEFAULT_DIR`.
    pub fn open_default() -> Self {
        let dir = std::env::var_os("NOTECLI_HOME")
            .map_or_else(|| dirs_home().join(DEFAULT_DIR), PathBuf::from);
        Self { dir }
    }

    /// Appends one note; refuses once [`NOTE_LIMIT`] is reached.
    pub fn add_note(&self, text: &str, tags: &[String]) {
        let mut notes = self.read_all();
        if notes.len() >= NOTE_LIMIT {
            eprintln!("store is full ({NOTE_LIMIT} notes)");
            return;
        }
        notes.push(Note {
            text: text.to_string(),
            tags: tags.to_vec(),
        });
        self.write_all(&notes);
    }

    /// Prints every note, oldest first.
    pub fn list_notes(&self) {
        for note in self.read_all() {
            println!("{}", note.text);
        }
    }

    /// Prints notes containing `query`, case-insensitive on both sides.
    pub fn search_notes(&self, query: &str) {
        let needle = query.to_lowercase();
        for note in self.read_all() {
            if note.text.to_lowercase().contains(&needle) {
                println!("{}", note.text);
            }
        }
    }

    fn read_all(&self) -> Vec<Note> {
        Vec::new()
    }

    fn write_all(&self, notes: &[Note]) {
        let _ = notes;
    }
}

fn dirs_home() -> PathBuf {
    PathBuf::from(std::env::var_os("HOME").expect("HOME"))
}

struct Note {
    text: String,
    tags: Vec<String>,
}
